File: src/me/kingtux/kingserverinfo/KingServerInfoMain.java
package me.kingtux.kingserverinfo;


import me.kingtux.kingserverinfo.bstats.Metrics;
import me.kingtux.kingserverinfo.commands.ServerInfoCommand;
import me.kingtux.kingserverinfo.events.ClickEvent;
import me.kingtux.kingserverinfo.events.JoinEvent;
import me.kingtux.kingserverinfo.events.LeaveEvent;
import me.kingtux.kingserverinfo.utils.MediaGui.MediaGui;
import me.kingtux.kingserverinfo.utils.config.ConfigManager;
import me.kingtux.kingserverinfo.utils.config.ConfigSettings;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.CommandMap;
import org.bukkit.plugin.java.JavaPlugin;

import java.lang.reflect.Field;
import java.util.logging.Level;

public class KingServerInfoMain extends JavaPlugin {
    private ConfigManager configManager;
    private ConfigSettings configSettings;
    private MediaGui mediaGui;

    public void onEnable() {
        if (Bukkit.getServer().getPluginManager().isPluginEnabled("PlaceholderAPI")) {
            configManager = new ConfigManager(this);
            configManager.setupConfig();

            getLogger().log(Level.INFO,
                    ChatColor.translateAlternateColorCodes('&',
                            configSettings.getPrefix() + " Successfully Got the Config and Settings"));
            try {
                final Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");

                bukkitCommandMap.setAccessible(true);
                CommandMap commandMap = (CommandMap) bukkitCommandMap.get(Bukkit.getServer());

                commandMap.register(configSettings.getServerInfoCommand(), new ServerInfoCommand(configSettings.getServerInfoCommand(), configSettings.getServerInfoDescription(), this));


            } catch (Exception e) {
                e.printStackTrace();
            }
            enableEvents();
            mediaGui = new MediaGui(this);
            Metrics metrics = new Metrics(this);


        } else {
            getLogger().log(Level.SEVERE, "I find your lack of the PlaceHolderAPI disturbing.");
        }



    }

    public void reloadPlugin() {
        configManager.reloadConfig();
        configSettings.getAllSettings();

        getLogger().log(Level.INFO,
                ChatColor.translateAlternateColorCodes('&',
                        configSettings.getPrefix() + " Successfully Got the Config and Settings"));

    }




    public void onDisable() {
        //Trying to clear out all used memory used by Configsettings
        System.gc();
    }

    public ConfigSettings getConfigSettings() {
        return configSettings;
    }

    public void setConfigSettings(ConfigSettings cF) {
        this.configSettings = cF;
    }

    private void enableEvents() {
        Bukkit.getServer().getPluginManager().registerEvents(new ClickEvent(this), this);
        Bukkit.getServer().getPluginManager().registerEvents(new LeaveEvent(this), this);
        Bukkit.getServer().getPluginManager().registerEvents(new JoinEvent(this), this);

    }


    public MediaGui getMediaGui() {
        return mediaGui;
    }
}

File: src/me/kingtux/kingserverinfo/KingServerInfoMain.kt
package me.kingtux.kingserverinfo


import me.kingtux.kingserverinfo.bstats.Metrics
import me.kingtux.kingserverinfo.commands.ServerInfoCommand
import me.kingtux.kingserverinfo.events.ClickEvent
import me.kingtux.kingserverinfo.events.JoinEvent
import me.kingtux.kingserverinfo.events.LeaveEvent
import me.kingtux.kingserverinfo.utils.config.ConfigManager
import me.kingtux.kingserverinfo.utils.config.ConfigSettings
import me.kingtux.kingserverinfo.utils.mediagui.MediaGui
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.CommandMap
import org.bukkit.plugin.java.JavaPlugin
import java.util.logging.Level

class KingServerInfoMain : JavaPlugin() {

    var configManager: ConfigManager? = null
    var configSettings: ConfigSettings? = null
    var mediaGui: MediaGui? = null
        private set

    override fun onEnable() {
        if (Bukkit.getServer().pluginManager.isPluginEnabled("PlaceholderAPI")) {
            configManager = ConfigManager(this)
            configManager!!.setupConfig()

            logger.log(Level.INFO,
                    ChatColor.translateAlternateColorCodes('&',
                            configSettings!!.prefix + " Successfully Got the Config and Settings"))
            try {
                val bukkitCommandMap = Bukkit.getServer().javaClass.getDeclaredField("commandMap")

                bukkitCommandMap.isAccessible = true
                val commandMap = bukkitCommandMap.get(Bukkit.getServer()) as CommandMap
                commandMap.register(configSettings!!.serverInfoCommand, ServerInfoCommand(configSettings!!.serverInfoCommand, configSettings!!.serverInfoDescription, this))

            } catch (e: Exception) {
                e.printStackTrace()
            }

            enableEvents()
            mediaGui = MediaGui(this)
            val metrics = Metrics(this)


        } else {
            logger.log(Level.SEVERE, "I find your lack of the PlaceHolderAPI disturbing.")
            server.pluginManager.disablePlugin(this)
        }


    }

    fun reloadPlugin() {
        configManager!!.reloadConfig()
        logger.log(Level.INFO,
                ChatColor.translateAlternateColorCodes('&',
                        configSettings!!.prefix + " Successfully Got the Config and Settings"))

    }


    override fun onDisable() {
    }

    private fun enableEvents() {
        Bukkit.getServer().pluginManager.registerEvents(ClickEvent(this), this)
        Bukkit.getServer().pluginManager.registerEvents(LeaveEvent(this), this)
        Bukkit.getServer().pluginManager.registerEvents(JoinEvent(this), this)

    }
}
